File: c/record.c
// Copyright 2019 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "record.h"

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "api.h"
#include "constants.h"

int is_block_type(byte typ) {
  switch (typ) {
    case BLOCK_TYPE_REF:
    case BLOCK_TYPE_LOG:
    case BLOCK_TYPE_OBJ:
    case BLOCK_TYPE_INDEX:
      return true;
  }
  return false;
}

int get_var_int(uint64_t *dest, struct slice in) {
  if (in.len == 0) {
    return -1;
  }

  int ptr = 0;
  uint64_t val = in.buf[ptr] & 0x7f;

  while (in.buf[ptr] & 0x80) {
    ptr++;
    if (ptr > in.len) {
      return -1;
    }
    val = (val + 1) << 7 | (uint64_t)(in.buf[ptr] & 0x7f);
  }

  *dest = val;
  return ptr + 1;
}

int put_var_int(struct slice dest, uint64_t val) {
  byte buf[10];

  int i = 9;
  buf[i] = (byte)(val & 0x7f);
  i--;
  while (true) {
    val >>= 7;
    if (!val) {
      break;
    }
    val--;
    buf[i] = 0x80 | (byte)(val & 0x7f);
    i--;
  }

  int n = sizeof(buf) - i - 1;
  if (dest.len < n) {
    return -1;
  }
  memcpy(dest.buf, &buf[i + 1], n);
  return n;
}

int common_prefix_size(struct slice a, struct slice b) {
  int p = 0;
  while (p < a.len && p < b.len) {
    if (a.buf[p] != b.buf[p]) {
      break;
    }
    p++;
  }

  return p;
}

int decode_string(struct slice *dest, struct slice in) {
  int start_len = in.len;
  uint64_t tsize = 0;
  int n = get_var_int(&tsize, in);
  if (n <= 0) {
    return -1;
  }
  in.buf += n;
  in.len -= n;
  if (in.len < tsize) {
    return -1;
  }

  slice_resize(dest, tsize + 1);
  dest->buf[tsize] = 0;
  memcpy(dest->buf, in.buf, tsize);
  in.buf += tsize;
  in.len -= tsize;

  return start_len - in.len;
}

int encode_key(bool *restart, struct slice dest, struct slice prev_key,
               struct slice key, byte extra) {
  struct slice start = dest;
  int prefix_len = common_prefix_size(prev_key, key);
  *restart = (prefix_len == 0);

  int n = put_var_int(dest, (uint64_t)prefix_len);
  if (n < 0) {
    return -1;
  }
  dest.buf += n;
  dest.len -= n;

  uint64_t suffix_len = key.len - prefix_len;
  n = put_var_int(dest, suffix_len << 3 | (uint64_t)extra);
  if (n < 0) {
    return -1;
  }
  dest.buf += n;
  dest.len -= n;

  if (dest.len < suffix_len) {
    return -1;
  }
  memcpy(dest.buf, key.buf + prefix_len, suffix_len);
  dest.buf += suffix_len;
  dest.len -= suffix_len;

  return start.len - dest.len;
}

byte ref_record_type() { return BLOCK_TYPE_REF; }

void ref_record_key(const void *r, struct slice *dest) {
  const struct ref_record *rec = (const struct ref_record *)r;
  slice_set_string(dest, rec->ref_name);
}

void ref_record_copy_from(void *rec, const void *src_rec) {
  struct ref_record *ref = (struct ref_record *)rec;
  struct ref_record *src = (struct ref_record *)src_rec;

  memset(ref, 0, sizeof(struct ref_record));
  if (src->ref_name != NULL) {
    ref->ref_name = strdup(src->ref_name);
  }

  if (src->target != NULL) {
    ref->target = strdup(src->target);
  }

  if (src->target_value != NULL) {
    ref->target_value = malloc(HASH_SIZE);
    memcpy(ref->target_value, src->target_value, HASH_SIZE);
  }

  if (src->value != NULL) {
    ref->value = malloc(HASH_SIZE);
    memcpy(ref->value, src->value, HASH_SIZE);
  }
  ref->update_index = src->update_index;
}

char hexdigit(int c) {
  if (c <= 9) {
    return '0' + c;
  }
  return 'a' + (c - 10);
}

void hex_format(char *dest, byte *src) {
  if (src != NULL) {
    for (int i = 0; i < HASH_SIZE; i++) {
      dest[2 * i] = hexdigit(src[i] >> 4);
      dest[2 * i + 1] = hexdigit(src[i] & 0xf);
    }
    dest[2 * HASH_SIZE] = 0;
  }
}

void ref_record_print(struct ref_record *ref) {
  char hex[41] = {};

  printf("ref{%s(%ld) ", ref->ref_name, ref->update_index);
  if (ref->value != NULL) {
    hex_format(hex, ref->value);
    printf("%s", hex);
  }
  if (ref->target_value != NULL) {
    hex_format(hex, ref->target_value);
    printf(" (T %s)", hex);
  }
  if (ref->target != NULL) {
    printf("=> %s", ref->target);
  }
  printf("}\n");
}

void ref_record_clear_void(void *rec) {
  ref_record_clear((struct ref_record *)rec);
}

void ref_record_clear(struct ref_record *ref) {
  free(ref->ref_name);
  free(ref->target);
  free(ref->target_value);
  free(ref->value);
  memset(ref, 0, sizeof(struct ref_record));
}

byte ref_record_val_type(const void *rec) {
  const struct ref_record *r = (const struct ref_record *)rec;
  if (r->value != NULL) {
    if (r->target_value != NULL) {
      return 2;
    } else {
      return 1;
    }
  } else if (r->target != NULL) {
    return 3;
  }
  return 0;
}

int ref_record_encode(const void *rec, struct slice s) {
  const struct ref_record *r = (const struct ref_record *)rec;
  struct slice start = s;
  int n = put_var_int(s, r->update_index);
  if (n < 0) {
    return -1;
  }
  s.buf += n;
  s.len -= n;

  if (r->value != NULL) {
    if (s.len < HASH_SIZE) {
      return -1;
    }
    memcpy(s.buf, r->value, HASH_SIZE);
    s.buf += HASH_SIZE;
    s.len -= HASH_SIZE;
  }

  if (r->target_value != NULL) {
    if (s.len < HASH_SIZE) {
      return -1;
    }
    memcpy(s.buf, r->target_value, HASH_SIZE);
    s.buf += HASH_SIZE;
    s.len -= HASH_SIZE;
  }

  if (r->target != NULL) {
    int l = strlen(r->target);
    n = put_var_int(s, l);
    if (n < 0) {
      return -1;
    }
    s.buf += n;
    s.len -= n;
    if (s.len < l) {
      return -1;
    }
    memcpy(s.buf, r->target, l);
    s.buf += l;
    s.len -= l;
  }

  return start.len - s.len;
}

int ref_record_decode(void *rec, struct slice key, byte val_type,
                      struct slice in) {
  struct ref_record *r = (struct ref_record *)rec;

  struct slice start = in;
  int n = get_var_int(&r->update_index, in);
  if (n < 0) {
    return n;
  }

  in.buf += n;
  in.len -= n;

  r->ref_name = realloc(r->ref_name, key.len + 1);
  memcpy(r->ref_name, key.buf, key.len);
  r->ref_name[key.len] = 0;
  switch (val_type) {
    case 1:
    case 2:
      if (in.len < HASH_SIZE) {
        return -1;
      }

      if (r->value == NULL) {
        r->value = malloc(HASH_SIZE);
      }
      memcpy(r->value, in.buf, HASH_SIZE);
      in.buf += HASH_SIZE;
      in.len -= HASH_SIZE;
      if (val_type == 1) {
        break;
      }
      if (r->target_value == NULL) {
        r->target_value = malloc(HASH_SIZE);
      }
      memcpy(r->target_value, in.buf, HASH_SIZE);
      in.buf += HASH_SIZE;
      in.len -= HASH_SIZE;
      break;
    case 3: {
      struct slice dest = {};
      int n = decode_string(&dest, in);
      if (n < 0) {
        return -1;
      }
      in.buf += n;
      in.len -= n;

      if (r->target != NULL) {
        free(r->target);
      }
      r->target = slice_to_string(dest);
      free(slice_yield(&dest));
    } break;

    case 0:
      break;
    default:
      abort();
      break;
  }

  return start.len - in.len;
}

int decode_key(struct slice *key, byte *extra, struct slice last_key,
               struct slice in) {
  int start_len = in.len;
  uint64_t prefix_len = 0;
  int n = get_var_int(&prefix_len, in);
  if (n < 0) {
    return -1;
  }
  in.buf += n;
  in.len -= n;

  if (prefix_len > last_key.len) {
    return -1;
  }

  uint64_t suffix_len = 0;
  n = get_var_int(&suffix_len, in);
  if (n <= 0) {
    return -1;
  }
  in.buf += n;
  in.len -= n;

  *extra = (byte)(suffix_len & 0x7);
  suffix_len >>= 3;

  if (in.len < suffix_len) {
    return -1;
  }

  slice_resize(key, suffix_len + prefix_len);
  memcpy(key->buf, last_key.buf, prefix_len);

  memcpy(key->buf + prefix_len, in.buf, suffix_len);
  in.buf += suffix_len;
  in.len -= suffix_len;

  return start_len - in.len;
}

struct record_ops ref_record_ops = {
    .key = &ref_record_key,
    .type = &ref_record_type,
    .copy_from = &ref_record_copy_from,
    .val_type = &ref_record_val_type,
    .encode = &ref_record_encode,
    .decode = &ref_record_decode,
    .clear = &ref_record_clear_void,
};

byte obj_record_type() { return BLOCK_TYPE_OBJ; }

void obj_record_key(const void *r, struct slice *dest) {
  const struct obj_record *rec = (const struct obj_record *)r;
  slice_resize(dest, rec->hash_prefix_len);
  memcpy(dest->buf, rec->hash_prefix, rec->hash_prefix_len);
}

void obj_record_copy_from(void *rec, const void *src_rec) {
  struct obj_record *ref = (struct obj_record *)rec;
  const struct obj_record *src = (const struct obj_record *)src_rec;

  *ref = *src;
  ref->hash_prefix = malloc(ref->hash_prefix_len);
  memcpy(ref->hash_prefix, src->hash_prefix, ref->hash_prefix_len);

  int olen = ref->offset_len * sizeof(uint64_t);
  ref->offsets = malloc(olen);
  memcpy(ref->offsets, src->offsets, olen);
}

void obj_record_clear(void *rec) {
  struct obj_record *ref = (struct obj_record *)rec;
  free(ref->hash_prefix);
  free(ref->offsets);
  memset(ref, 0, sizeof(struct obj_record));
}

byte obj_record_val_type(const void *rec) {
  struct obj_record *r = (struct obj_record *)rec;
  if (r->offset_len > 0 && r->offset_len < 8) {
    return r->offset_len;
  }
  return 0;
}

int obj_record_encode(const void *rec, struct slice s) {
  struct obj_record *r = (struct obj_record *)rec;
  struct slice start = s;
  if (r->offset_len == 0 || r->offset_len >= 8) {
    int n = put_var_int(s, r->offset_len);
    if (n < 0) {
      return -1;
    }
    s.buf += n;
    s.len -= n;
  }
  if (r->offset_len == 0) {
    return start.len - s.len;
  }
  int n = put_var_int(s, r->offsets[0]);
  if (n < 0) {
    return -1;
  }
  s.buf += n;
  s.len -= n;

  uint64_t last = r->offsets[0];
  for (int i = 1; i < r->offset_len; i++) {
    int n = put_var_int(s, r->offsets[i] - last);
    if (n < 0) {
      return -1;
    }
    s.buf += n;
    s.len -= n;
    last = r->offsets[i];
  }

  return start.len - s.len;
}

int obj_record_decode(void *rec, struct slice key, byte val_type,
                      struct slice in) {
  struct slice start = in;
  struct obj_record *r = (struct obj_record *)rec;

  r->hash_prefix = malloc(key.len);
  memcpy(r->hash_prefix, key.buf, key.len);
  r->hash_prefix_len = key.len;

  uint64_t count = val_type;
  if (val_type == 0) {
    int n = get_var_int(&count, in);
    if (n < 0) {
      return n;
    }

    in.buf += n;
    in.len -= n;
  }

  r->offsets = NULL;
  r->offset_len = 0;
  if (count == 0) {
    return start.len - in.len;
  }

  r->offsets = malloc(count * sizeof(uint64_t));
  r->offset_len = count;

  int n = get_var_int(&r->offsets[0], in);
  if (n < 0) {
    return n;
  }

  in.buf += n;
  in.len -= n;

  uint64_t last = r->offsets[0];

  int j = 1;
  while (j < count) {
    uint64_t delta = 0;
    int n = get_var_int(&delta, in);
    if (n < 0) {
      return n;
    }

    in.buf += n;
    in.len -= n;

    last = r->offsets[j] = (delta + last);
    j++;
  }
  return start.len - in.len;
}

struct record_ops obj_record_ops = {
    .key = &obj_record_key,
    .type = &obj_record_type,
    .copy_from = &obj_record_copy_from,
    .val_type = &obj_record_val_type,
    .encode = &obj_record_encode,
    .decode = &obj_record_decode,
    .clear = &obj_record_clear,
};

struct record new_record(byte typ) {
  struct record rec;
  switch (typ) {
    case BLOCK_TYPE_REF: {
      struct ref_record *r = calloc(1, sizeof(struct ref_record));
      record_from_ref(&rec, r);
      return rec;
    }

    case BLOCK_TYPE_OBJ: {
      struct obj_record *r = calloc(1, sizeof(struct obj_record));
      record_from_obj(&rec, r);
      return rec;
    }

    case BLOCK_TYPE_INDEX: {
      struct index_record *r = calloc(1, sizeof(struct index_record));
      record_from_index(&rec, r);
      return rec;
    }
  }
  abort();
  return rec;
}

byte index_record_type() { return BLOCK_TYPE_INDEX; }

void index_record_key(const void *r, struct slice *dest) {
  struct index_record *rec = (struct index_record *)r;
  slice_copy(dest, rec->last_key);
}

void index_record_copy_from(void *rec, const void *src_rec) {
  struct index_record *dst = (struct index_record *)rec;
  struct index_record *src = (struct index_record *)src_rec;

  slice_copy(&dst->last_key, src->last_key);
  dst->offset = src->offset;
}

void index_record_clear(void *rec) {
  struct index_record *idx = (struct index_record *)rec;
  free(slice_yield(&idx->last_key));
}

byte index_record_val_type(const void *rec) { return 0; }

int index_record_encode(const void *rec, struct slice out) {
  const struct index_record *r = (const struct index_record *)rec;
  struct slice start = out;

  int n = put_var_int(out, r->offset);
  if (n < 0) {
    return n;
  }

  out.buf += n;
  out.len -= n;

  return start.len - out.len;
}

int index_record_decode(void *rec, struct slice key, byte val_type,
                        struct slice in) {
  struct slice start = in;
  struct index_record *r = (struct index_record *)rec;

  slice_copy(&r->last_key, key);

  int n = get_var_int(&r->offset, in);
  if (n < 0) {
    return n;
  }

  in.buf += n;
  in.len -= n;
  return start.len - in.len;
}

struct record_ops index_record_ops = {
    .key = &index_record_key,
    .type = &index_record_type,
    .copy_from = &index_record_copy_from,
    .val_type = &index_record_val_type,
    .encode = &index_record_encode,
    .decode = &index_record_decode,
    .clear = &index_record_clear,
};

void record_key(struct record rec, struct slice *dest) {
  rec.ops->key(rec.data, dest);
}

byte record_type(struct record rec) { return rec.ops->type(); }

int record_encode(struct record rec, struct slice dest) {
  return rec.ops->encode(rec.data, dest);
}

void record_copy_from(struct record rec, struct record src) {
  assert(src.ops->type() == rec.ops->type());

  rec.ops->copy_from(rec.data, src.data);
}

byte record_val_type(struct record rec) { return rec.ops->val_type(rec.data); }

int record_decode(struct record rec, struct slice key, byte extra,
                  struct slice src) {
  return rec.ops->decode(rec.data, key, extra, src);
}

void record_clear(struct record rec) { return rec.ops->clear(rec.data); }

void record_from_ref(struct record *rec, struct ref_record *ref_rec) {
  rec->data = ref_rec;
  rec->ops = &ref_record_ops;
}

void record_from_obj(struct record *rec, struct obj_record *obj_rec) {
  rec->data = obj_rec;
  rec->ops = &obj_record_ops;
}

void record_from_index(struct record *rec, struct index_record *index_rec) {
  rec->data = index_rec;
  rec->ops = &index_record_ops;
}

void *record_yield(struct record *rec) {
  void *p = rec->data;
  rec->data = NULL;
  return p;
}

struct ref_record *record_as_ref(struct record rec) {
  assert(record_type(rec) == BLOCK_TYPE_REF);
  return (struct ref_record *)rec.data;
}

bool hash_equal(byte *a, byte *b) {
  if (a != NULL && b != NULL) {
    return 0 == memcmp(a, b, HASH_SIZE);
  }

  return a == b;
}

bool str_equal(char *a, char *b) {
  if (a != NULL && b != NULL) {
    return 0 == strcmp(a, b);
  }

  return a == b;
}

bool ref_record_equal(struct ref_record *a, struct ref_record *b) {
  return 0 == strcmp(a->ref_name, b->ref_name) &&
         a->update_index == b->update_index && hash_equal(a->value, b->value) &&
         hash_equal(a->target_value, b->target_value) &&
         str_equal(a->target, b->target);
}

bool ref_record_is_deletion(const struct ref_record *ref) {
  return ref->value == NULL && ref->target == NULL && ref->target_value == NULL;
}

File: record.go
/*
Copyright 2020 Google LLC

Use of this source code is governed by a BSD-style
license that can be found in the LICENSE file or at
https://developers.google.com/open-source/licenses/bsd
*/

package reftable

import (
	"encoding/binary"
	"errors"
	"fmt"
	"math"
)

func newRecord(typ byte, key string) record {
	switch typ {
	case blockTypeLog:
		l := &LogRecord{}
		if key != "" {
			if !l.decodeKey(key) {
				panic("LogRecord key should be well formed")
			}
		}
		return l
	case blockTypeRef:
		return &RefRecord{
			RefName: key,
		}
	case blockTypeObj:
		return &objRecord{
			HashPrefix: []byte(key),
		}
	case blockTypeIndex:
		return &indexRecord{LastKey: key}
	}
	return nil
}

type objRecord struct {
	HashPrefix []byte
	Offsets    []uint64
}

func getVarInt(buf []byte) (uint64, int) {
	if len(buf) == 0 {
		return 0, -1
	}
	ptr := 0
	val := uint64(buf[ptr] & 0x7f)
	for buf[ptr]&0x80 != 0 {
		ptr++
		if ptr > len(buf) {
			return 0, -1
		}
		val = ((val + 1) << 7) | uint64(buf[ptr]&0x7f)
	}

	return val, ptr + 1
}

func putVarInt(buf []byte, val uint64) (n int, ok bool) {
	var dest [10]byte

	i := 9
	dest[i] = byte(val & 0x7f)
	i--
	for {
		val >>= 7
		if val == 0 {
			break
		}

		val--
		dest[i] = 0x80 | byte(val&0x7f)
		i--
	}

	n = len(dest[i+1:])
	if n > len(buf) {
		return 0, false
	}
	copy(buf, dest[i+1:])
	return n, true
}

func encodeKey(buf []byte, prevKey, key string, extra uint8) (n int, restart bool, fits bool) {
	start := buf
	prefixLen := commonPrefixSize(prevKey, key)
	restart = prefixLen == 0

	s, ok := putVarInt(buf, uint64(prefixLen))
	if !ok {
		return
	}
	buf = buf[s:]

	suffixLen := len(key) - prefixLen
	if s, ok := putVarInt(buf, uint64(suffixLen<<3)|uint64(extra)); !ok {
		return
	} else {
		buf = buf[s:]
	}
	if len(buf) < suffixLen {
		return
	}
	copy(buf, key[prefixLen:])
	buf = buf[suffixLen:]

	n = len(start) - len(buf)
	return n, restart, true
}

func (r *RefRecord) typ() byte {
	return blockTypeRef
}

func (r *RefRecord) key() string {
	return r.RefName
}

func (r *RefRecord) copyFrom(in record) {
	*r = *in.(*RefRecord)
}

func (r *RefRecord) isDeletion() bool {
	return r.Value == nil && r.TargetValue == nil && r.Target == ""
}

func (r *RefRecord) String() string {
	return fmt.Sprintf("ref(%s)", r.RefName)
}

func (r *RefRecord) decode(buf []byte, key string, valType uint8, hashSize int) (n int, ok bool) {
	*r = RefRecord{}
	start := buf
	r.RefName = key
	delta, s := getVarInt(buf)
	if s <= 0 {
		return
	}
	r.UpdateIndex = delta
	buf = buf[s:]

	switch valType {
	case 1, 2:
		if len(buf) < hashSize {
			return
		}
		r.Value = make([]byte, hashSize)
		copy(r.Value, buf[:hashSize])
		buf = buf[hashSize:]
		if valType == 1 {
			break
		}
		if len(buf) < hashSize {
			return
		}
		r.TargetValue = make([]byte, hashSize)
		copy(r.TargetValue, buf[:hashSize])
		buf = buf[hashSize:]
	case 3:
		tsize, s := getVarInt(buf)
		if s <= 0 {
			return
		}
		buf = buf[s:]
		if len(buf) < int(tsize) {
			return
		}

		r.Target = string(buf[:tsize])
		buf = buf[tsize:]
	}

	return len(start) - len(buf), true
}

func (r *RefRecord) valType() uint8 {
	var valueType uint8
	if len(r.Value) > 0 {
		if len(r.TargetValue) > 0 {
			valueType = 2
		} else {
			valueType = 1
		}
	} else if len(r.Target) > 0 {
		valueType = 3
	}
	return valueType
}

func (r *RefRecord) encode(buf []byte, hashSize int) (n int, fits bool) {
	start := buf

	s, ok := putVarInt(buf, uint64(r.UpdateIndex))
	if !ok {
		return
	}
	buf = buf[s:]

	if s := len(r.Value); s > 0 {
		if len(buf) < s {
			return
		}
		copy(buf, r.Value)
		buf = buf[s:]
	}
	if s := len(r.TargetValue); s > 0 {
		if len(buf) < s {
			return
		}
		copy(buf, r.TargetValue)
		buf = buf[s:]
	}
	if len(r.Target) > 0 {
		s, ok := putVarInt(buf, uint64(len(r.Target)))
		if !ok {
			return
		}
		buf = buf[s:]
		if len(buf) < len(r.Target) {
			return
		}
		copy(buf, r.Target)
		buf = buf[len(r.Target):]
	}
	return len(start) - len(buf), true
}

func (r *objRecord) key() string {
	return string(r.HashPrefix)
}

func (r *objRecord) copyFrom(in record) {
	*r = *in.(*objRecord)
}

func (r *objRecord) typ() byte {
	return blockTypeObj
}

func (r *objRecord) String() string {
	return fmt.Sprintf("obj(%x)", r.HashPrefix)
}

func (r *objRecord) valType() uint8 {
	var lower uint8
	if l := len(r.Offsets); l > 0 && l < 8 {
		lower = uint8(l)
	}

	return lower
}

func (r *objRecord) encode(buf []byte, hashSize int) (n int, fits bool) {
	start := buf

	if len(r.Offsets) == 0 || len(r.Offsets) >= 8 {
		s, ok := putVarInt(buf, uint64(len(r.Offsets)))
		if !ok {
			return
		}
		buf = buf[s:]
	}

	if len(r.Offsets) == 0 {
		return len(start) - len(buf), true
	}
	s, ok := putVarInt(buf, uint64(r.Offsets[0]))
	if !ok {
		return
	}
	buf = buf[s:]

	last := r.Offsets[0]
	for _, o := range r.Offsets[1:] {
		s, ok := putVarInt(buf, o-last)
		if !ok {
			return
		}
		buf = buf[s:]
		last = o
	}

	return len(start) - len(buf), true
}

func (r *objRecord) decode(buf []byte, prefix string, cnt3 uint8, hashSize int) (n int, ok bool) {
	*r = objRecord{}

	start := buf
	r.HashPrefix = []byte(prefix)
	var count uint64
	if cnt3 == 0 {
		count, n = getVarInt(buf)
		if n <= 0 {
			return
		}
		buf = buf[n:]
	} else {
		count = uint64(cnt3)
	}

	if count == 0 {
		return len(start) - len(buf), true
	}

	r.Offsets = make([]uint64, 1, count)
	r.Offsets[0], n = getVarInt(buf)
	if n <= 0 {
		return
	}
	buf = buf[n:]
	count--

	last := r.Offsets[0]
	for count > 0 {
		o, n := getVarInt(buf)
		if n <= 0 {
			return 0, false
		}
		buf = buf[n:]
		count--

		o += last
		r.Offsets = append(r.Offsets, o)
		last = o
	}
	return len(start) - len(buf), true
}

type indexRecord struct {
	LastKey string
	Offset  uint64
}

func (r *indexRecord) key() string {
	return r.LastKey
}

func (r *indexRecord) typ() byte {
	return blockTypeIndex
}

func (r *indexRecord) valType() byte {
	return 0
}

func (r *indexRecord) copyFrom(in record) {
	*r = *in.(*indexRecord)
}

func (r *indexRecord) String() string {
	return fmt.Sprintf("idx(%s)", r.LastKey)
}

func (r *indexRecord) decode(buf []byte, key string, valType uint8, hashSize int) (n int, ok bool) {
	*r = indexRecord{}
	start := buf
	r.LastKey = key

	var s int
	r.Offset, s = getVarInt(buf)
	if s <= 0 {
		return
	}
	buf = buf[s:]
	return len(start) - len(buf), true
}

func (r *indexRecord) encode(buf []byte, hashSize int) (n int, ok bool) {
	start := buf

	s, ok := putVarInt(buf, uint64(r.Offset))
	if !ok {
		return
	}
	buf = buf[s:]
	return len(start) - len(buf), true
}

var fmtError = errors.New("reftable: format error")

func decodeRestartKey(buf []byte, off uint32) (key string, err error) {
	err = fmtError
	if len(buf) <= int(off) {
		return
	}

	if buf[off] != 0 {
		return
	}
	buf = buf[off+1:]

	l, s := getVarInt(buf)
	if s <= 0 {
		return
	}
	buf = buf[s:]
	l >>= 3
	if uint64(len(buf)) < l {
		return
	}

	return string(buf[:l]), nil
}

func decodeKey(buf []byte, prevKey string) (n int, key string, value uint8, ok bool) {
	start := buf
	prefixLen, s := getVarInt(buf)
	if s <= 0 {
		return
	}
	buf = buf[s:]

	suffixLen, s := getVarInt(buf)
	if s <= 0 {
		return
	}
	buf = buf[s:]
	value = uint8(suffixLen & 0x7)
	suffixLen = suffixLen >> 3

	if int(suffixLen) > len(buf) {
		return
	}

	if int(prefixLen) > len(prevKey) {
		return
	}

	name := make([]byte, suffixLen+prefixLen)
	copy(name, prevKey[:prefixLen])
	copy(name[prefixLen:], buf[:suffixLen])
	buf = buf[suffixLen:]
	return len(start) - len(buf), string(name), value, true
}

func revInt64(t uint64) uint64 {
	return math.MaxUint64 - t
}

func encodeString(buf []byte, val string) (n int, ok bool) {
	start := buf
	if s, ok := putVarInt(buf, uint64(len(val))); !ok {
		return 0, false
	} else {
		buf = buf[s:]
	}
	if len(buf) < len(val) {
		return
	}
	buf = buf[copy(buf, val):]

	return len(start) - len(buf), true
}

func (l *LogRecord) typ() byte {
	return blockTypeLog
}

func (l *LogRecord) key() string {
	var suffix [9]byte
	binary.BigEndian.PutUint64(suffix[1:], revInt64(l.UpdateIndex))
	return l.RefName + string(suffix[:])
}

func (r *LogRecord) copyFrom(in record) {
	*r = *in.(*LogRecord)
}

func (l *LogRecord) String() string {
	return fmt.Sprintf("log(%s, %d)", l.RefName, l.UpdateIndex)
}

func (l *LogRecord) valType() uint8 {
	return 0x1
}

func (l *LogRecord) encode(buf []byte, hashSize int) (n int, fits bool) {
	if l.Old == nil {
		l.Old = make([]byte, hashSize)
	}
	if l.New == nil {
		l.New = make([]byte, hashSize)
	}

	if len(l.Old) != hashSize || len(l.New) != hashSize {
		panic("invalid log entry")
	}

	start := buf

	if len(buf) < len(l.Old) {
		return
	}
	buf = buf[copy(buf, l.Old):]

	if len(buf) < len(l.New) {
		return
	}
	buf = buf[copy(buf, l.New):]

	if s, ok := encodeString(buf, l.Name); !ok {
		return
	} else {
		buf = buf[s:]
	}
	if s, ok := encodeString(buf, l.Email); !ok {
		return
	} else {
		buf = buf[s:]
	}

	if s, ok := putVarInt(buf, l.Time); !ok {
		return
	} else {
		buf = buf[s:]
	}
	if len(buf) < 2 {
		return
	}
	binary.BigEndian.PutUint16(buf, uint16(l.TZOffset))
	buf = buf[2:]

	if s, ok := encodeString(buf, l.Message); !ok {
		return
	} else {
		buf = buf[s:]
	}
	return len(start) - len(buf), true
}

func decodeString(buf []byte) (n int, val string, ok bool) {
	start := buf
	nameLen, s := getVarInt(buf)
	if s <= 0 {
		return
	}
	buf = buf[s:]
	if len(buf) < int(nameLen) {
		return
	}
	val = string(buf[:nameLen])
	buf = buf[nameLen:]
	return len(start) - len(buf), val, true
}

func (l *LogRecord) decodeKey(key string) bool {
	if len(key) < 10 {
		return false
	}
	l.RefName = key[:len(key)-9]
	last := []byte(key[len(key)-9:])
	if last[0] != 0 {
		return false
	}
	l.UpdateIndex = revInt64(binary.BigEndian.Uint64(last[1:]))
	return true
}

func (l *LogRecord) decode(buf []byte, key string, valType uint8, hashSize int) (n int, ok bool) {
	*l = LogRecord{}
	start := buf

	if !l.decodeKey(key) {
		return
	}

	if valType == 0 {
		return 0, true
	}
	buf = buf[n:]

	if len(buf) < 2*hashSize {
		return
	}

	l.Old = make([]byte, hashSize)
	l.New = make([]byte, hashSize)
	copy(l.Old, buf[:hashSize])
	buf = buf[hashSize:]
	copy(l.New, buf[:hashSize])
	buf = buf[hashSize:]

	n, l.Name, ok = decodeString(buf)
	if !ok {
		return
	}
	buf = buf[n:]

	n, l.Email, ok = decodeString(buf)
	if !ok {
		return
	}
	buf = buf[n:]

	l.Time, n = getVarInt(buf)
	if n <= 0 {
		return
	}
	buf = buf[n:]

	if len(buf) < 2 {
		return
	}
	tz := binary.BigEndian.Uint16(buf)
	buf = buf[2:]
	l.TZOffset = int16(tz)

	n, l.Message, ok = decodeString(buf)
	if !ok {
		return
	}
	buf = buf[n:]
	return len(start) - len(buf), true
}
